# Задача объяснения смысла

### Вспомогательные функции

freeDictionary API https://github.com/maxzhenzhera/python-freeDictionaryAPI

In [70]:
!pip install python-freeDictionaryAPI httpx

In [90]:
from freedictionaryapi.clients.sync_client import DictionaryApiClient

def get_definitions(search_query, top_k_results=5):
    with DictionaryApiClient() as client:
         parser = client.fetch_parser(search_query)        
    definitions = []
    for meaning in parser.word.meanings:
        for definition in meaning.definitions:
            definitions.append(definition.definition)
    return definitions[:top_k_results]

Wikipedia API

In [45]:
!pip install wikipedia-api

In [63]:
import wikipediaapi
import requests

API_KEY = 'YOUR_APP_NAME (YOUR_EMAIL_OR_CONTACT_PAGE)'

def get_wikipedia_summary(search_query, top_k_results=5, language_code='en'):
    headers = {
        'User-Agent': API_KEY
    }

    base_url = 'https://api.wikimedia.org/core/v1/wikipedia/'
    endpoint = '/search/page'
    url = base_url + language_code + endpoint
    parameters = {'q': search_query, 'limit': top_k_results}
    response = requests.get(url, headers=headers, params=parameters)
    
    titles = [page['title'] for page in response.json()['pages']]
    
    wiki_wiki = wikipediaapi.Wikipedia(
    user_agent=API_KEY,
        language=language_code,
        extract_format=wikipediaapi.ExtractFormat.WIKI
    )

    summaries = [wiki_wiki.page(title).summary for title in titles]
    return summaries

Функция для создания нового промпта

In [107]:
def make_prompt(source_prompt, source_word, top_k_definitions=3, top_k_wikipedia=2):
    promt = f'Here are some ground truth information about {source_word}:\n'
    summaries = get_wikipedia_summary(source_word, top_k_results=top_k_wikipedia)
    definitions = get_definitions(source_word, top_k_results=top_k_definitions)
    if len(summaries) > 0:
        promt += ''.join(summaries) + '\n'
    if len(definitions) > 0:
        promt += ''.join(definitions) + '\n'  
        
    return promt[:-2] + '.\nUsing provided evidence above, ' + source_prompt

In [112]:
print(make_prompt('Can you explain what is billon?', 'billon', top_k_definitions=3, top_k_wikipedia=2))

Here are some ground truth information about billon:
Billon may refer to:

Billon (alloy), a metal alloy containing mostly copper or bronze with small quantity of silverBillon () is an alloy of a precious metal (most commonly silver, but also gold) with a majority base metal content (such as copper). It is used chiefly for making coins, medals, and token coins.
The word comes from the French bille, which means "log".
Any of several alloys of precious metals and base metals that are used (where legal) to make coins, medals etc.
Using provided evidence above, Can you explain what is billon?


### Агоритм починки

**Шаг 1 - Написание промпта**.

Q: Can you explain what is billon?

A: Sure, a billion is a number equal to 1,000 million or 1,000,000,000 (one thousand million). It has 9 zeros after the initial digit...

**Шаг 2 - Поиск релевантных знаний в интернете и создание соответствующего промпта**


Шаг предполагает использование написанной функции `make_prompt(source_prompt, source_word, top_k_definitions=3, top_k_wikipedia=2)`

Q: Here are some ground truth information about billon:
Billon may refer to:

Billon (alloy), a metal alloy containing mostly copper or bronze with small quantity of silverBillon () is an alloy of a precious metal (most commonly silver, but also gold) with a majority base metal content (such as copper). It is used chiefly for making coins, medals, and token coins.
The word comes from the French bille, which means "log".
Any of several alloys of precious metals and base metals that are used (where legal) to make coins, medals etc.
Using provided evidence above, Can you explain what is billon?

A: Certainly! According to the provided information, "billon" refers to an alloy used in the production of coins, medals, and token coins. This alloy is typically composed of a precious metal, most commonly silver but sometimes gold, with a majority base metal content such as copper. The purpose of using billon is to create a material that has some of the desirable properties of the precious metal while being more cost-effective due to the inclusion of a less expensive base metal.

The term "billon" is derived from the French word "bille," which means "log." It is likely that this term was historically associated with the alloy due to its composition or appearance.

In summary, billon is a type of alloy that combines a precious metal with a base metal, primarily used in the minting of coins, medals, and token coins.

### Chatgpt fix

#### 1. Billon

Q: Can you explain what is billon?

A: Certainly! The term ""billion"" can have different meanings depending on the context and the region in which it's used. Historically, there have been variations in the definition of a billion, but in contemporary usage, there are two main systems: the long scale and the short scale.

1. Long scale: In some European countries, such as France and Germany, the long scale system was traditionally used. In this system, a billion equals a million millions, which is equivalent to 1,000,000,000,000 (1 followed by 12 zeros). This system extends similarly with trillion being 1 followed by 18 zeros, quadrillion as 1 followed by 24 zeros, and so forth.

2. Short scale: The short scale system is commonly used in English-speaking countries and is also recognized by various international organizations. In this system, a billion equals a thousand millions, which is equivalent to 1,000,000,000 (1 followed by 9 zeros). The short scale continues with a trillion being 1 followed by 12 zeros, quadrillion as 1 followed by 15 zeros, and so on.

Due to the differences in these definitions, the term ""billion"" can lead to confusion when dealing with very large numbers. To mitigate this confusion, it's sometimes helpful to specify whether you're using the long scale or the short scale when referring to extremely large numerical values.

Q: Here are some ground truth information about billon:
Billon may refer to:

Billon (alloy), a metal alloy containing mostly copper or bronze with small quantity of silverBillon () is an alloy of a precious metal (most commonly silver, but also gold) with a majority base metal content (such as copper). It is used chiefly for making coins, medals, and token coins.
The word comes from the French bille, which means "log".
Any of several alloys of precious metals and base metals that are used (where legal) to make coins, medals etc.
Using provided evidence above, Can you explain what is billon?

A: Certainly! According to the provided information, "billon" refers to an alloy used in the production of coins, medals, and token coins. This alloy is typically composed of a precious metal, most commonly silver but sometimes gold, with a majority base metal content such as copper. The purpose of using billon is to create a material that has some of the desirable properties of the precious metal while being more cost-effective due to the inclusion of a less expensive base metal.

The term "billon" is derived from the French word "bille," which means "log." It is likely that this term was historically associated with the alloy due to its composition or appearance.

In summary, billon is a type of alloy that combines a precious metal with a base metal, primarily used in the minting of coins, medals, and token coins.

#### 2. Mousehole

Q: Can you explain what is mousehole?

A: "In the context of oil and gas drilling, a ""mousehole"" refers to a secondary hole that's drilled adjacent to the primary drilling hole, known as the ""rat hole."" These holes are typically found on drilling rigs.

Here's a breakdown of these terms:

1. **Rat Hole:** This is the initial, shallower hole drilled to set the surface casing. Once this hole reaches a certain depth, it's set aside and covered (usually with a protective casing) to maintain its integrity. This allows the main drilling operation to continue without obstruction.

2. **Mousehole:** The mousehole is a smaller diameter hole drilled close to the primary wellbore. It's often used for storing the drill pipe that's not actively being used during the drilling process. This hole is smaller and shallower than the primary wellbore.

The purpose of the mousehole is to provide a convenient location to temporarily store sections of the drill string when they're not in use. This helps in efficiently managing the drilling process, allowing for quick access to additional sections of the drill string when needed without taking up valuable space on the drilling rig floor."


In [113]:
print(make_prompt('Can you explain what is mousehole?', 'mousehole', top_k_definitions=3, top_k_wikipedia=1))

Here are some ground truth information about mousehole:
Mousehole (; Cornish: Porthenys) is a village and fishing port in Cornwall, England, UK. It is approximately 2.5 miles (4 km) south of Penzance on the shore of Mount's Bay. The village is in the civil parish of Penzance. An islet called St Clement's Isle lies about 350 metres (380 yd) offshore from the harbour entrance.
Mousehole lies within the Cornwall Area of Outstanding Natural Beauty (AONB). 27% of Cornwall has AONB designation, with the same status and protection as a National Park.
A hole through which a mouse enters a room.Any small hole or opening.(drilling) The storage area on a drilling rig where the next joint of drilling pipe is held until needed.
Using provided evidence above, Can you explain what is mousehole?


A: Based on the provided information, "Mousehole" can refer to multiple things:

Location in Cornwall, England:

Mousehole is a village and fishing port in Cornwall, England, located approximately 2.5 miles south of Penzance on the shore of Mount's Bay.
It is situated in the civil parish of Penzance.
There is an islet called St Clement's Isle, located about 350 meters offshore from the harbor entrance.
Geographical and Environmental Context:

Mousehole is within the Cornwall Area of Outstanding Natural Beauty (AONB), indicating that the area is recognized for its natural and scenic significance.
The Cornwall AONB designation provides the same status and protection as a National Park, emphasizing the importance of preserving the area's natural beauty.
Alternative Definitions:

In a general sense, a "mousehole" refers to a hole through which a mouse enters a room, or any small hole or opening.
In the context of drilling, it can also refer to the storage area on a drilling rig where the next joint of drilling pipe is held until needed.
It's important to note that the term "mousehole" can have different meanings depending on the context in which it is used. In the geographical context provided, it primarily refers to the village and fishing port in Cornwall, England.

### Claude fix

Q: Can you explain what is billon?
A: 
A billion is a number represented as either:

1,000,000,000 (10 to the power of 9) - This is known as the short scale definition used in countries like the United States and Canada.
OR

1,000,000,000,000 (10 to the power of 12) - This is known as the long scale definition used in countries like the United Kingdom and France.
The difference comes down to how various numeric scales are defined.

In short scale countries, the scales goes:
Thousand, Million, Billion, Trillion, Quadrillion, Quintillion, etc...
Where each term is 1,000 times greater than the previous one.

In long scale countries, the scale goes:
Thousand, Million, Milliard, Billion, Billiard, Trillion, Trilliard, etc...
Where each term is 1,000,000 times greater than the previous one.

So a billion equals either:

1,000 millions (short scale) or
1 million millions (long scale)
This numeric terminology difference can cause confusion when comparing large numbers internationally. Care must be taken to clarify which scale is being referred to when using terms like billion.

Q: Here are some ground truth information about billon: Billon may refer to:

Billon (alloy), a metal alloy containing mostly copper or bronze with small quantity of silverBillon () is an alloy of a precious metal (most commonly silver, but also gold) with a majority base metal content (such as copper). It is used chiefly for making coins, medals, and token coins. The word comes from the French bille, which means "log". Any of several alloys of precious metals and base metals that are used (where legal) to make coins, medals etc. Using provided evidence above, Can you explain what is billon?

A: After reviewing the additional context you provided, I understand that billon refers specifically to a metal alloy containing mostly copper or bronze, with a small quantity of silver or sometimes gold. More specifically:

Billon is an alloy used chiefly for making coins, medals, and token coins. It contains a precious metal like silver or gold combined with a base metal like copper or bronze, with the base metal making up the majority of the content.

The name "billon" comes from the French word "bille" meaning log or ingot. This refers to the fact that coins and medals were originally cast from logs or ingots of this billon alloy.

So in summary, billon is not a pure precious metal, but rather an alloy containing a small amount of silver or gold mixed with mostly copper or bronze. It has properties suitable for minting and casting purposes like coinage, while using less of the costly precious metals. The specific ratios and metals can vary, but the key defining traits are:

Primarily copper/bronze base metal
Smaller quantity of silver or gold
Used for making coins, medals, tokens

### Gigachad fix

Q: Can you explain what is billon?

A: "Billon" is a term that can have different meanings depending on the context. In general, it refers to a unit of measurement or a value representing a very large number.
In mathematics, a "billion" is typically used as a synonym for the number 1,000,000,000 (one thousand million). It is often used in place of the word "trillion" when referring to smaller numbers.
In some countries, "billion" is used as a prefix in the metric system to denote a factor of 10^9 (ten to the ninth power). For example, a "billionaire" is someone who has a net worth of one billion dollars.
In financial and economic contexts, "billion" is often used to describe large sums of money or values. For instance, a company might report billion-dollar profits or a country's GDP might exceed a billion dollars.
It's important to note that the use of "billion" can vary across different countries and languages. In some places, it may refer to a different value or unit of measurement.

Q: Here are some ground truth information about billon: Billon may refer to:

Billon (alloy), a metal alloy containing mostly copper or bronze with small quantity of silverBillon () is an alloy of a precious metal (most commonly silver, but also gold) with a majority base metal content (such as copper). It is used chiefly for making coins, medals, and token coins. The word comes from the French bille, which means "log". Any of several alloys of precious metals and base metals that are used (where legal) to make coins, medals etc. Using provided evidence above, Can you explain what is billon?

A: 
Based on the information provided, "billon" can refer to different things. In the context of an alloy, "billon" is an alloy of a precious metal (usually silver) with a majority base metal content (such as copper). It is commonly used to make coins, medals, and token coins.
In terms of unit of measurement, "billion" is typically used as a synonym for the number 1,000,000,000 (one thousand million). It is often used in place of the word "trillion" when referring to smaller numbers.
It's important to note that the use of "billion" can vary across different countries and languages. In some places, it may refer to a different value or unit of measurement.

In [114]:
print(make_prompt('Can you explain what is billon?', 'billon', top_k_definitions=3, top_k_wikipedia=2))

Here are some ground truth information about billon:
Billon may refer to:

Billon (alloy), a metal alloy containing mostly copper or bronze with small quantity of silverBillon () is an alloy of a precious metal (most commonly silver, but also gold) with a majority base metal content (such as copper). It is used chiefly for making coins, medals, and token coins.
The word comes from the French bille, which means "log".
Any of several alloys of precious metals and base metals that are used (where legal) to make coins, medals etc.
Using provided evidence above, Can you explain what is billon?


И так далее... Гигачад не чинится